core-processor/src/main/java/io/micronaut/inject/writer/DispatchWriter.java

```diff
@@ -514,19 +526,57 @@ public TypedElement getDeclaringType() {
 
         @Override
         public ExpressionDef dispatchExpression(ExpressionDef bean) {
-            final TypeDef propertyType = TypeDef.of(beanField.getType());
-            final ClassTypeDef targetType = ClassTypeDef.of(beanField.getOwningType());
-
-            if (beanField.isReflectionRequired()) {
-                return TYPE_REFLECTION_UTILS.invokeStatic(
-                    METHOD_GET_FIELD_VALUE,
-                    ExpressionDef.constant(targetType), // Target class
-                    ExpressionDef.constant(beanField.getName()), // Field name,
-                    bean // Target instance
-                ).cast(propertyType);
-            } else {
-                return bean.cast(targetType).field(beanField).cast(propertyType);
-            }
+            return bean.cast(ClassTypeDef.of(beanField.getOwningType()))
+                .field(beanField)
+                .cast(TypeDef.of(beanField.getType()));
+        }
+
+        @NonNull
+        public FieldElement getField() {
+            return beanField;
+        }
+    }
+
+    /**
+     * Field get dispatch target.
+     */
+    @Internal
+    public static final class FieldGetReflectionDispatchTarget extends AbstractDispatchTarget {
+        @NonNull
+        final FieldElement beanField;
+
+        public FieldGetReflectionDispatchTarget(FieldElement beanField) {
+            this.beanField = beanField;
+        }
+
+        @Override
+        public boolean supportsDispatchOne() {
+            return true;
+        }
+
+        @Override
+        public boolean supportsDispatchMulti() {
+            return false;
+        }
+
+        @Override
+        public MethodElement getMethodElement() {
+            return null;
+        }
+
+        @Override
+        public TypedElement getDeclaringType() {
+            return null;
+        }
+
+        @Override
+        public ExpressionDef dispatchExpression(ExpressionDef bean) {
+            return TYPE_REFLECTION_UTILS.invokeStatic(
```

Comment:
Would be nice to move this to method handles eventually. Might even work with native image, then